lib.php

```
@@ -295,6 +295,20 @@ function kialo_update_discussion_url(int $coursemoduleid, string $updateddiscuss
 
     $result = $DB->update_record('kialo', $updatedata);
     if (!$result) {
-        throw new dml_exception('updaterecordfailed', 'Failed to update kialo record');
+        throw new \moodle_exception('errors:updaterecordfailed', 'kialo');
```

I wanted to test this case but couldn't find a way to trigger it.

I tried these two diffs:

(1)

```diff
diff --git i/lib.php w/lib.php
index 87d8992..0e4a086 100644
--- i/lib.php
+++ w/lib.php
@@ -289,7 +289,7 @@ function kialo_update_discussion_url(int $coursemoduleid, string $updateddiscuss
 
     // Update the discussion_url field in the kialo table.
     $updatedata = new stdClass();
-    $updatedata->id = $kialoid;
+    $updatedata->id = 9999999;
     $updatedata->discussion_url = $updateddiscussionurl;
     $updatedata->timemodified = time();
 
```

Result: No error, backend logs "Successfully updated discussion_url for resource_link".

(2)

```diff
diff --git i/lib.php w/lib.php
index 87d8992..234da2c 100644
--- i/lib.php
+++ w/lib.php
@@ -291,7 +291,7 @@ function kialo_update_discussion_url(int $coursemoduleid, string $updateddiscuss
     $updatedata = new stdClass();
     $updatedata->id = $kialoid;
     $updatedata->discussion_url = $updateddiscussionurl;
-    $updatedata->timemodified = time();
+    $updatedata->timemodified = "not a date";
 
     $result = $DB->update_record('kialo', $updatedata);
     if (!$result) {
```

Result: This triggers an error, but not `'errors:updaterecordfailed'`. The backend logs:

`[error    ] Failed to update target URL    [kialo.operations.lti.lti] controller=LtiResourceLinkController http_status=500 method=GET previous_url=https://localhost.kialolabs.com:5000/13024 request_id= response='{"error":"Error writing to database (Incorrect integer value: \'not a date\' for column `moodle`.`mdl_kialo`.`timemodified` at row 1\\nUPDATE mdl_kialo SET discussion_url = ?,timemodified = ? WHERE id=?\\n[array (\\n  0 => \'https:\\/\\/localhost.kialolabs.com:5000\\/13028\',\\n  1 => \'not a date\',\\n  2 => \'2\',\\n)])"}' sentry=True trace_id=54137f49-ee7c-4be8-8187-bbb66d57a4db url=https://localhost.kialolabs.com:5000/13028`

Do you know a way to trigger this `'errors:updaterecordfailed'`? I realise you didn't add this case in this PR. Maybe @oha-kialo you know?

---

I've always just "mocked" what was happening by always throwing the error. I would've expected it to go wrong in the cases you listed, but the documentation also doesn't really say when it would return false https://moodledev.io/docs/5.1/apis/core/dml#update_record

Maybe `DML exceptions: New DML code is throwing exceptions instead of returning false if anything goes wrong` at the bottom of this document is an indicator that we're not actually ever returning false?

---

Yeah I did the same, to test I just commented out the line `$result = $DB->update_record...` so that `$result` is always null.

---

> Maybe `DML exceptions: New DML code is throwing exceptions instead of returning false if anything goes wrong` at the bottom of this document is an indicator that we're not actually ever returning false?

Ah yeah that could be an explanation.